### PR TITLE
Loading Changed and Fixed

### PR DESCRIPTION
**Pull request by vsinh14**
# Description

This fixes the saving and loading functionality for both player and maps team, by fixing some bugs and making the code quality better so that the loading and saving system is more understandable. Furthermore major fixes have been made to the load button implementation in the menu and the functionality of the load button.

Fixes / Closes # (issue)
https://github.com/UQcsse3200/2024-studio-1/issues/406
#417 
#420 
#421 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Feature Improvement 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Menu components were implementation tested through multiple 
Unit tests for map and character saving 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java:1]: Do we really need a whole new screen just for a loaded game, all g if you do just wondering

Comment (vsinh14) [reply source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java:1]: u mean file wise yeah it keeps the code clear

**Comment by vsinh14** [comment]
Fixed issue with loading pets from save files

File: source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java
Comment (IanP96) [reply source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java:1]: this adds a lot of code duplication though, surely you could have a base GameScreen class or something, or just another instance of MainGameScreen with different parameters

Comment (vsinh14) [reply source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java:1]: Done

**Comment by AElanagai** [comment]
Looks Good , some methods will have to be moved to gamearea from maingamearea like displayui() as refactoring is done that way but that could be done during merge too.

**Review by AElanagai — APPROVED** [review]

File: source/core/src/main/com/csse3200/game/screens/MainGameScreen.java [hunk 1]
```diff
@@ -126,30 +66,40 @@ public MainGameScreen(GdxGame game) {
...
+     * is later set to completed in the create method above
+     * @param filePath: Path for the save file
+     */
+    public MapLoadConfig loadFromJson(String filePath) {
+        MapLoadConfig mapLoadConfig = new MapLoadConfig();
+        mapLoadConfig  = FileLoader.readClass(MapLoadConfig.class, filePath, FileLoader.Location.EXTERNAL);
+        return mapLoadConfig;
+    }
+
+    private Entity loadPlayer(boolean shouldLoad, GameOptions gameOptions, String chosenPlayer, Difficulty difficulty) {
```
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/screens/MainGameScreen.java:102]: doubt that you need difficulty here, you're already passing in gameOptions

**Comment by IanP96** [comment]
`GameOptions.shouldLoad` is never meaningfully used (never read from) now that the load button just loads immediately. Don't think you'll be using that anymore, you can probably delete it

File: source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java [hunk 2]
```diff
@@ -50,12 +51,13 @@ public void create() {
...
-        ArrayList mapLoadConfig = FileLoader.readClass(ArrayList.class, MainGameArea.MAP_SAVE_PATH, FileLoader.Location.EXTERNAL);
-        PlayerConfig playerConfig = FileLoader.readClass(PlayerConfig.class, "configs/player_save.json", FileLoader.Location.EXTERNAL);
+        MapLoadConfig mapLoadConfig = FileLoader.readClass(MapLoadConfig.class,
+                MainGameArea.MAP_SAVE_PATH, FileLoader.Location.EXTERNAL);
+        PlayerConfig playerConfig = FileLoader.readClass(PlayerConfig.class,
+                "saves/player_save.json", FileLoader.Location.EXTERNAL);
 
-        logger.info("{}\n{}\n{}", playerConfig, mapLoadConfig, playerLocationConfig);
-        return playerLocationConfig != null && mapLoadConfig != null && playerConfig != null;
+        logger.info("{}\n{}\n{}", playerConfig, mapLoadConfig);
```
[the remark this answers is not in the document]
Comment (IanP96) [reply source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java:59]: the info message I mean

File: source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java [hunk 3]
```diff
@@ -148,15 +156,20 @@ public void changed(ChangeEvent changeEvent, Actor actor) {
...
+            for (TextButton button : new TextButton[]{
+                    howToPlayBtn, achievementsBtn, settingsBtn, exitBtn}) {
+                table.add(button);
+                table.row();
+            }
 
-        for (TextButton button : new TextButton[]{
-                howToPlayBtn, achievementsBtn, settingsBtn, exitBtn}) {
-            table.add(button);
-            table.row();
```
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java:159]: why is this changed? You can just keep the original code but replace `shouldLoadBtn` with `loadButton`. Then you don't need two for loops

**Comment by IanP96** [comment]
> `GameOptions.shouldLoad` is never meaningfully used (never read from) now that the load button just loads immediately. Don't think you'll be using that anymore, you can probably delete it

You can delete the boolean param from MainMenuActions.onPlayerSelect and the event invocation, if you're going to the player select screen in the first place then you're definitely not loading

File: source/core/src/main/com/csse3200/game/screens/MainGameScreen.java [hunk 4]
```diff
@@ -126,30 +66,40 @@ public MainGameScreen(GdxGame game) {
...
+    /**
+     * Loads map data from the save file and extracts the completed rooms into a list which
+     * is later set to completed in the create method above
+     * @param filePath: Path for the save file
+     */
+    public MapLoadConfig loadFromJson(String filePath) {
+        MapLoadConfig mapLoadConfig = new MapLoadConfig();
+        mapLoadConfig  = FileLoader.readClass(MapLoadConfig.class, filePath, FileLoader.Location.EXTERNAL);
+        return mapLoadConfig;
+    }
```
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/screens/MainGameScreen.java:100]: this is unused?

**Review by IanP96 — CHANGES_REQUESTED** [review]
Mostly looks good, just made some suggestions to simplify things but no functionality changes

File: source/core/src/main/com/csse3200/game/components/npc/attack/MeleeAttackComponent.java
Comment (mblauberg) [inline source/core/src/main/com/csse3200/game/components/npc/attack/MeleeAttackComponent.java:1]: super class already checks for null target in update loop before running performAttack

**Review by mblauberg — APPROVED** [review]

File: source/core/src/main/com/csse3200/game/components/npc/attack/MeleeAttackComponent.java
Comment (vsinh14) [reply source/core/src/main/com/csse3200/game/components/npc/attack/MeleeAttackComponent.java:1]: it was crashing on going to a new room and attacking an entity that was removed needed to be done for the bug ticket as listed in the pr

File: source/core/src/main/com/csse3200/game/screens/MainGameScreen.java (hunk 1 above)
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/screens/MainGameScreen.java:102]: resolved!

File: source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java (hunk 3 above)
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java:159]: changed it!

File: source/core/src/main/com/csse3200/game/screens/MainGameScreen.java (hunk 4 above)
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/screens/MainGameScreen.java]: deleted it.. we had it just in case if we needed to refer back to it but we changed the implementation now.

File: source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (manya-k) [reply source/core/src/main/com/csse3200/game/components/mainmenu/MainMenuDisplay.java]: done

**Comment by vsinh14** [comment]
> `GameOptions.shouldLoad` is never meaningfully used (never read from) now that the load button just loads immediately. Don't think you'll be using that anymore, you can probably delete it

if you have not saved a game locally, you will not see the load button unless that has changed recently

File: source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java [hunk 5]
```diff
@@ -1,34 +1,129 @@
...
+    private final PlayerFactory playerFactory;
+    private static final String[] mainGameTextures = {
+            "images/heart.png",
+            "images/ui_white_icons.png",
+            "images/ui_white_icons_over.png",
+            "images/ui_white_icons_down.png",
+            "skins/rainbow/skin/rainbow-ui.png",
+            "images/dot_transparent.png",
+            "images/black_dot_transparent.png"
+    };
```
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/screens/LoadGameScreen.java:59]: duplicated in MainGameScreen as well, you could put this in GameScreen

**Review by IanP96 — APPROVED** [review]
lgtm just some code duplication but otherwise all looks v good
